### PR TITLE
Do not crash on invalid index access

### LINKED ISSUE #5057: Rational number cannot be represented as fixed point type causes crash
Found by solfuzzer.

```
contract C {
    function f() public returns (string memory) {
        return (["somethingShort", "0123456789012345678901^P34@678789012345<FF>^?89012567890123456789023123456789012345345678901234567890123456789001234567890123456789012345#78901234567890123456789012345%70123/567890122345%70123/567890123456785788900000000000000000000000000000000089012345676666666666666666666666662345678901234567894123456789012(4567890"][9012345678900123123456789012345678912345678901234<37890123456-89012345678901234567890123456789012345678901234557890123456789012345678900123123456789012345345678901234567890123456789001234567890123456789012345678901234567890123456789012345%70123/567890122345%70123/567890123456785788900000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001234567890123456789012][9012345678900123123456789012345678912300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000045678901234<37890123456-89012300000000000000000045678901234567890123456789012345678901234557890123456789012345678900123123456789012345345123456789001234567890123456789567890123456789012345678901234567890123456789012345%70123/567890122345%70123/56789000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001234567890123456789012][90123456789001231234567890123456789123000000000000000000000000000000000123456785788900000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000345670123/56789012345678901231]);
    }
}
```

Error:
```
Internal compiler error during compilation:
/Users/alex/Projects-Changes-new/solidity/libsolidity/ast/Types.cpp(1279): Throw in function virtual dev::u256 dev::solidity::RationalNumberType::literalValue(const dev::solidity::Literal *) const
Dynamic exception type: boost::exception_detail::clone_impl<dev::solidity::InternalCompilerError>
std::exception::what: 
[dev::tag_comment*] = 
```

Also uploaded two examples as a file, because it contains unprintable characters.

[id:000313,sig:06,src:015830,op:havoc,rep:4.txt](https://github.com/ethereum/solidity/files/2405923/id.000313.sig.06.src.015830.op.havoc.rep.4.txt)

[id:000485,sig:06,src:018965,op:havoc,rep:4.txt](https://github.com/ethereum/solidity/files/2405952/id.000485.sig.06.src.018965.op.havoc.rep.4.txt)

### LINKED ISSUE #5058: Too large multi-dimensional array causes crash
Found by solfuzzer.

```
contract C {
  function f(bytes32[1263941234127518272][500] memory) public pure {}
}
```

Error:
```
Internal compiler error during compilation:
/Users/alex/Projects/solidity/libsolidity/ast/Types.cpp(1722): Throw in function virtual unsigned int dev::solidity::ArrayType::calldataEncodedSize(bool) const
Dynamic exception type: boost::exception_detail::clone_impl<dev::solidity::InternalCompilerError>
std::exception::what: Array size does not fit unsigned.
[dev::tag_comment*] = Array size does not fit unsigned.
```

### LINKED ISSUE #5059: Invalid number literal causes crash
Found by solfuzzer.

```
contract C {
    function f() public returns (string memory) {
        return (["somethingShort", "01234567890123456ESC8901^P34@6783456789001234567890123453789012345678901234567890123456789012^E<FF><FF>^E78901234590123455789012345678901234567890012312345678901234567890123?567890123456789012345678901234567890123456745678901234567890123456789012345678901234567890123456789012345%%%%%%%%67890123456#8901234567890"][9001234567891234567890123453789012345>-.90123456789<1234567890123888888888888888888888888888888888888888888888888888888888888-888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888-888888888888888885888888888888888888888888888888888888888888888^88888888888888^8888888888][90123678901234567891555555555555523456789012336789012345678915555555555555234567890123453789012345&888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888-888888888888888888888888888888888888888888888888888888888888888^888888888888^8888888885555888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888789012325678955555555555555555555555555555555555555555555555555555555555555555555558888888888888888888888888888888888888888888888888888888^8888555555555555555555588888][901236789012345678915555555555555234567890123453789012345&88888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888-888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888-888888888888888885888888888888888888888888888888888888888888888^88888888888888^8888888888][901236789012345678915555555555555234567890123453789012345&888888888888888888888888888888888888888888888555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555523453789012345&-891234567890123453789012345&-8901234567890123256789555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555501234&67890123457778888888888^8888888888][908888888888888888888888888888888888888888888^8888555555555555555555588888][901236789012345678915555555555555234567890123453789012345&88888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888-888888888888888888888888888888888888888888888888888888888888888^888888888888^8888888885555888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888789012325678955555555555588888888888888888888888888888888888888888888888888888^88885555555555555555555555555555555555555558-888888888888888885888888888888888888888888888888888888888888888^88888888888888^8888888888][901236789012345678915555555555555234567890123453789012345&8888888888888888888888888888888888888888888888888888888>888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888-888888888888888888888888888888888888888888888888888888888888888^888888888888^8888888885555888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888789012325678955555555555555555555555555555555555555555555555555555555555555555555558888888888888888888888888888888888888888888888888888888^8888555555555555555555588888][901236789012345678915555555555555234567890123453789012345&88888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888-8888888888888888888888888888888888888888888808888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888887890123256789555555555555555555555555555555555555555555555555555555555<5555555555558888888888888888888888888888888888888888888888888888888^8888555555555555555555555555555555555555555555588^8/888888555588^8888888885555888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888789012325678955555555555555555555555555555555555555552555555555555555555555555555558888888888888888888888888888888888888888888888888888888^888855555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555238888888888888888888888888888888888888888888888888888888888888888888-88888888888888855555555555555555555555555555555555555555555555555555555555888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888^888888888888888888888888888888888888888888888-888888888888888888888888888888888888888888888888888888888888888^88888888888855555558888^888888888888888888888888888888888888888888888-88888888888888855555555588888888888888888888888888888888888888-888888888888888888888888888888888888<888888888888888888^888888888888888888888888888888888888888888888-888888888888888888888888888888888888888888888888888888888888888^8888888888885555555555555588888888888889888888888888888888&8888888888888888888888888888888888888888888888888888888888888888888888888888888888789012325678955555555555555555555555555555555555555555555555555555555555555555555558888888888888888888888888888888888888888888888888888888^85555555555555555555555555555555555|555555555555555555555555555555555555555555555555555555555555555555555555555555555555555558888888555555555555555|55555555555588888888888888888888888888888888888888888888881234567891234567890123453789012345&-891234567890123453789012345&-89012345678901232567895555555555555555555555555555555555555555555|555555555555555555555555555555555555555555555555555555555555555555|55555555555555555555588888888888888888^8888555555555555555555588888888888888888888888888888888888888888888888888^888888888888888888888888888888888888888888888888888888888888888888888888-88888888888888855555555555555555555555555555555555555555555555555555555555888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888^888888888888888888888888888888888888888888888-888888888888888888888888888888888888888888888888888888888888888^888888888888555555555555558888888888888988888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888789012325678955555555555555555555555555555555555555555555555555555555555555555555558888888888888888888888888888888888888888888888888888888^888888888888888888888888888888888888888888888-888888888888888555555555888888888888888888855555555588888888888888888888888888-8888<888555588888888888888888888888888888888888888888888888888^888888888888888888888888888888888888888888888-8888888888888885555555555555555555555555555555555555555555555555555555555555555888888888888888888888888888888888^888888888888888888888888888888888888888888888-888888888888888888888888888888888888<888888888888888888^888888888888888888888888888888888888888888888-888888888888888888888888888888888888888888888888888888888888888^888888888888555555555555558888888888888988888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888789012325678955555555555555555555555555555555555555555555555555555555555555555555558888888888888888888888888888888888888888888888888888888^888888888888888888888888888888888888888888888-88888888888888855555555588888888888888888855555555555555555555555555555555888888888888888888888888888888888888888888888888888881234567891234567890123453789012345&-891234567890123453789012345&-89012345678901232567895555555555555555555555555555555555555555555|5555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555501234&6789012345777777777777777]);
    }
}
```

Error
```
Internal compiler error during compilation:
/Users/alex/Projects/solidity/libsolidity/ast/Types.cpp(1286): Throw in function virtual dev::u256 dev::solidity::RationalNumberType::literalValue(const dev::solidity::Literal *) const
Dynamic exception type: boost::exception_detail::clone_impl<dev::solidity::InternalCompilerError>
std::exception::what: Number constant too small.
[dev::tag_comment*] = Number constant too small.
```

Also uploaded as a file, because it contains unprintable characters.

[id:000427,sig:06,src:006817+019026,op:splice,rep:2.txt](https://github.com/ethereum/solidity/files/2405946/id.000427.sig.06.src.006817%2B019026.op.splice.rep.2.txt)

Note: this is in the same codebase but a different assertion than #5056.

### PR DESCRIPTION
**Pull request by axic**
Fixes #5057.
Depends on https://github.com/ethereum/solidity/pull/5195/.

Another instance of this https://github.com/ethereum/solidity/blob/develop/libsolidity/analysis/TypeChecker.cpp#L2253 uses `hasErrors()`.

I think that is unreliable and we may just use `fatalError` in `expectType`.

### CODE REVIEW

**Comment by axic** [comment]
Need to add tests.

File: test/libsolidity/syntaxTests/indexing/array_multidim_overflow_index.sol [hunk 1]
```diff
@@ -0,0 +1,6 @@
+contract C {
+  function f() public {
+    bytes[32] memory a;
+    a[8**90][8**90][1 - 8**90];
+  }
+}
```
Comment (axic) [inline test/libsolidity/syntaxTests/indexing/array_multidim_overflow_index.sol]: This is failing with an assertion, to be fixed by this PR.

File: test/libsolidity/syntaxTests/indexing/array_multidim_rational.sol [hunk 2]
```diff
@@ -0,0 +1,6 @@
+contract C {
+  function f() public {
+    bytes[32] memory a;
+    a[8**90][8**90][8**90*0.1];
+  }
+}
```
Comment (axic) [inline test/libsolidity/syntaxTests/indexing/array_multidim_rational.sol]: This is failing with an assertion, to be fixed by this PR.

File: test/libsolidity/syntaxTests/array/length/parameter_too_large_multidim.sol [hunk 3]
```diff
@@ -0,0 +1,3 @@
+contract C {
+  function f(bytes32[1263941234127518272][500] memory) public pure {}
+}
```
Comment (axic) [inline test/libsolidity/syntaxTests/array/length/parameter_too_large_multidim.sol]: This is failing with an assertion, to be fixed by this PR.

File: test/libsolidity/syntaxTests/indexing/array_multidim_rational.sol (hunk 2 above)
Comment (axic) [inline test/libsolidity/syntaxTests/indexing/array_multidim_rational.sol]: This is failing with an assertion, to be fixed by this PR.

File: test/libsolidity/syntaxTests/indexing/fixedbytes_noninteger_index.sol [hunk 4]
```diff
@@ -0,0 +1,6 @@
+contract C {
+  function f() public {
+    bytes32 b;
+    b[888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888];
+  }
+}
```
Comment (axic) [inline test/libsolidity/syntaxTests/indexing/fixedbytes_noninteger_index.sol]: This is failing with an assertion, to be fixed by this PR.

File: libsolidity/analysis/TypeChecker.cpp [hunk 5]
```diff
@@ -2280,15 +2280,17 @@ bool TypeChecker::visit(IndexAccess const& _access)
...
+			if (!expectType(*index, IntegerType(256)) || type(*index)->category() != Type::Category::RationalNumber)
 				m_errorReporter.fatalTypeError(index->location(), "Integer constant expected.");
+
+			auto length = dynamic_cast<RationalNumberType const*>(type(*index).get());
+			solAssert(length, "");
+
+			resultType = make_shared<TypeType>(make_shared<ArrayType>(
+				DataLocation::Memory,
+				typeType.actualType(),
+				length->literalValue(nullptr)
```
Comment (axic) [inline libsolidity/analysis/TypeChecker.cpp:2292]: I think we are missing another test case here for too large rational numbers.

**Comment by axic** [comment]
@chriseth please review this, the last failure is due to the need of checking `calldatasize` in the type checker for a multi dimensional array and rejecting if it is too large.

File: libsolidity/analysis/TypeChecker.cpp [hunk 6]
```diff
@@ -2280,15 +2280,17 @@ bool TypeChecker::visit(IndexAccess const& _access)
...
 		{
-			expectType(*index, IntegerType(256));
-			if (auto length = dynamic_cast<RationalNumberType const*>(type(*index).get()))
-				resultType = make_shared<TypeType>(make_shared<ArrayType>(
-					DataLocation::Memory,
-					typeType.actualType(),
-					length->literalValue(nullptr)
-				));
-			else
+			if (!expectType(*index, IntegerType(256)) || type(*index)->category() != Type::Category::RationalNumber)
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp:2283]: expectType already creates an error message

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp:2283]: Ok, yeah, it was like that before, but then if we make `expectType` return a bool, should not the condition be negated?

**Review by chriseth — DISMISSED** [review]

File: test/libsolidity/syntaxTests/array/length/parameter_too_large_multidim.sol (hunk 3 above)
Comment (erak) [reply test/libsolidity/syntaxTests/array/length/parameter_too_large_multidim.sol]: Done.

File: test/libsolidity/syntaxTests/indexing/array_multidim_overflow_index.sol (hunk 1 above)
Comment (erak) [reply test/libsolidity/syntaxTests/indexing/array_multidim_overflow_index.sol]: Done.

File: test/libsolidity/syntaxTests/indexing/fixedbytes_noninteger_index.sol (hunk 4 above)
Comment (erak) [reply test/libsolidity/syntaxTests/indexing/fixedbytes_noninteger_index.sol]: Done.

File: test/libsolidity/syntaxTests/indexing/array_multidim_rational.sol (hunk 2 above)
Comment (erak) [reply test/libsolidity/syntaxTests/indexing/array_multidim_rational.sol]: Done.

File: libsolidity/analysis/TypeChecker.cpp (hunk 5 above)
Comment (erak) [reply libsolidity/analysis/TypeChecker.cpp:2292]: This should be already covered by e.g. ``array_multidim_rational.sol``. Is that correct or do you have another test in mind?

File: libsolidity/analysis/TypeChecker.cpp [hunk 7]
```diff
@@ -822,12 +822,12 @@ bool TypeChecker::visit(VariableDeclaration const& _variable)
...
 	case Type::Category::Array:
 		if (auto arrayType = dynamic_cast<ArrayType const*>(varType.get()))
-			if (
-				((arrayType->location() == DataLocation::Memory) ||
-				(arrayType->location() == DataLocation::CallData)) &&
-				!arrayType->validForCalldata()
-			)
-				m_errorReporter.typeError(_variable.location(), "Array is too large to be encoded.");
+		{
+			auto baseType = dynamic_cast<ArrayType const*>(arrayType->baseType().get());
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Why can you assume that the base type is also an array?

File: libsolidity/analysis/TypeChecker.cpp [hunk 8]
```diff
@@ -822,12 +822,12 @@ bool TypeChecker::visit(VariableDeclaration const& _variable)
...
+			if (arrayType->location() == DataLocation::Memory || arrayType->location() == DataLocation::CallData)
+				if ((baseType && !baseType->validForCalldata()) || !arrayType->validForCalldata())
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: If the base type is not valid for calldata, why should the type itself be? This rather sounds like the change should be inside `validForCalldata`

**Review by chriseth — CHANGES_REQUESTED** [review]
This PR needs to be split into multiple PRs.

**Comment by erak** [comment]
I split out https://github.com/ethereum/solidity/pull/5188 and https://github.com/ethereum/solidity/pull/5195. Since these PRs are very related to each other, I based my decision on where to split them on the parts of the compiler they change and not on the test cases they cover. 

This PR now depends on https://github.com/ethereum/solidity/pull/5195, because the new signature of ``expectType()`` will only be introduced by that one.

**Comment by bit-shift** [comment]
After checking again, it became clear that #5195 already fixed the bug that this PR is trying to fix. Will wait for @axic to verify before closing this.

**Comment by erak** [comment]
After checking again, it became clear that #5195 already fixed the remaining bug that this PR is trying to fix. I will close it now as this change does not seem to be needed anymore.

**Pull request closed without merging**